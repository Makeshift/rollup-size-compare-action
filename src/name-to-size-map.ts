import type {Sizes} from './types'

function findAllChildren(start: any = {}): any {
  return start.children ? start.children.flatMap(findAllChildren) : [start]
}

function trimPath(input: string): string {
  if (!input) {
    return ''
  }

  return input.replace(/.*node_modules/, '/node_modules')
}

export function assetNameToSizeMap(statAssets: any = {}): Map<string, Sizes> {
  return new Map(
    statAssets?.tree?.children.map((asset: any) => {
      const children = findAllChildren(asset)
      let size = 0
      let gzipSize: number | null = statAssets.options.gzip ? 0 : null
      for (const mod of children) {
        size += statAssets.nodeParts[mod.uid].renderedLength
        if (statAssets.options.gzip) {
          gzipSize += statAssets.nodeParts[mod.uid].gzipLengh
        }
      }

      return [
        trimPath(asset.name),
        {
          size,
          gzipSize
        }
      ]
    })
  )
}

export function chunkModuleNameToSizeMap(
  statChunks: any = {}
): Map<string, Sizes> {
  if (!statChunks?.tree) {
    return new Map()
  }
  return new Map(
    findAllChildren(statChunks.tree).map((mod: any) => {
      const modInfo = statChunks.nodeParts[mod.uid]
      return [
        trimPath(statChunks.nodeMetas[modInfo.metaUid].id),
        {
          size: modInfo.renderedLength ?? 0,
          gzipSize: statChunks.options.gzip ? modInfo.gzipLengh : null
        }
      ]
    })
  )
}
