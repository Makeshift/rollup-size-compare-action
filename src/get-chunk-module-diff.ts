import {chunkModuleNameToSizeMap} from './name-to-size-map'
import type {WebpackStatsDiff} from './types'
import {webpackStatsDiff} from './webpack-stats-diff'

export function getChunkModuleDiff(
  oldStats: any,
  newStats: any
): WebpackStatsDiff | null {
  return webpackStatsDiff(
    chunkModuleNameToSizeMap(oldStats),
    chunkModuleNameToSizeMap(newStats)
  )
}
