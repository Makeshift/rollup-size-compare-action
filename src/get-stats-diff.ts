import {assetNameToSizeMap} from './name-to-size-map'
import type {WebpackStatsDiff} from './types'
import {webpackStatsDiff} from './webpack-stats-diff'

export function getStatsDiff(
  oldAssetStats: any,
  newAssetStats: any
): WebpackStatsDiff {
  return webpackStatsDiff(
    assetNameToSizeMap(oldAssetStats),
    assetNameToSizeMap(newAssetStats)
  )
}
