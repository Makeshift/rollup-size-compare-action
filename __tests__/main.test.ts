import {expect, test} from '@jest/globals'
import {getStatsDiff} from '../src/get-stats-diff'
import {getChunkModuleDiff} from '../src/get-chunk-module-diff'
import {
  printAssetTablesByGroup,
  printChunkModulesTable,
  printTotalAssetTable
} from '../src/print-markdown'
import {AssetDiff} from '../src/types'
import {readFile} from 'node:fs/promises'
import {resolve} from 'node:path'

async function readJsonFile(path: string): Promise<any> {
  const data = await readFile(resolve(__dirname, path), 'utf8')
  return JSON.parse(data)
}

test('Shows stats when files are removed', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/old-stats-rollup-sample.json'),
    await readJsonFile('./__mocks__/new-stats-rollup-sample.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('Shows stats when files are added', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/new-stats-rollup-sample.json'),
    await readJsonFile('./__mocks__/old-stats-rollup-sample.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('Shows stats when files are unchanged', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/old-stats-rollup-sample.json'),
    await readJsonFile('./__mocks__/old-stats-rollup-sample.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('computes the correct module diff information', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-rollup.json'),
    await readJsonFile('./__mocks__/new-stats-rollup.json')
  )

  expect(statsDiff?.added).toContainEqual({
    name: '/node_modules/actions/src/client/this-file-was-added.ts',
    diff: 1496,
    diffPercentage: Infinity,
    new: {size: 1496, gzipSize: NaN},
    old: {size: 0, gzipSize: 0}
  } as AssetDiff)
  expect(statsDiff?.bigger).toContainEqual({
    name: '/node_modules/actions/src/client/this-file-grew-larger.tsx',
    diff: 200,
    diffPercentage: 35.58719,
    new: {size: 762, gzipSize: NaN},
    old: {size: 562, gzipSize: NaN}
  } as AssetDiff)
  expect(statsDiff?.smaller).toContainEqual({
    name: '/node_modules/actions/src/client/helpers/this-file-grew-smaller.ts',
    diff: -200,
    diffPercentage: -7.94281,
    new: {size: 2318, gzipSize: NaN},
    old: {size: 2518, gzipSize: NaN}
  } as AssetDiff)
  expect(statsDiff?.removed).toContainEqual({
    name: '/node_modules/actions/src/client/this-file-will-be-deleted.ts',
    diff: -1496,
    diffPercentage: -100,
    new: {size: 0, gzipSize: 0},
    old: {size: 1496, gzipSize: NaN}
  } as AssetDiff)
  expect(statsDiff?.total.new).toEqual(statsDiff?.total.old)
  expect(statsDiff?.total.diff).toEqual(0)
  expect(statsDiff?.total.diffPercentage).toEqual(0)
})

test('displays module information when files are added/removed/changed', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-rollup.json'),
    await readJsonFile('./__mocks__/new-stats-rollup.json')
  )

  expect(printChunkModulesTable(statsDiff)).toMatchSnapshot()
})

test('displays no module information when unchanged', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-rollup.json'),
    await readJsonFile('./__mocks__/old-stats-rollup.json')
  )

  expect(printChunkModulesTable(statsDiff)).toMatchSnapshot()
})
