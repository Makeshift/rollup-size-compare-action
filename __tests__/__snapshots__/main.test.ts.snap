// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows stats when files are added 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
1 â†’ 2 | 122.29 kB â†’ 140.66 kB (+18.37 kB) | +15.02%"
`;

exports[`Shows stats when files are added 2`] = `
"**Added**

Asset | File Size | % Changed
----- | --------- | ---------
static/js/resize-observer-6FaCDon1.js | 0 B â†’ 18.37 kB (+18.37 kB) | -

**Removed**

No assets were removed

**Bigger**

No assets were bigger

**Smaller**

No assets were smaller

**Unchanged**

Asset | File Size | % Changed
----- | --------- | ---------
static/js/BackgroundImage-hmXOL_KY.js | 122.29 kB | 0%"
`;

exports[`Shows stats when files are removed 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
2 â†’ 1 | 140.66 kB â†’ 122.29 kB (-18.37 kB) | -13.06%"
`;

exports[`Shows stats when files are removed 2`] = `
"**Added**

No assets were added

**Removed**

Asset | File Size | % Changed
----- | --------- | ---------
static/js/resize-observer-6FaCDon1.js | 18.37 kB â†’ 0 B (-18.37 kB) | -100%

**Bigger**

No assets were bigger

**Smaller**

No assets were smaller

**Unchanged**

Asset | File Size | % Changed
----- | --------- | ---------
static/js/BackgroundImage-hmXOL_KY.js | 122.29 kB | 0%"
`;

exports[`Shows stats when files are unchanged 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
2 | 140.66 kB | 0%"
`;

exports[`Shows stats when files are unchanged 2`] = `
"**Added**

No assets were added

**Removed**

No assets were removed

**Bigger**

No assets were bigger

**Smaller**

No assets were smaller

**Unchanged**

Asset | File Size | % Changed
----- | --------- | ---------
static/js/resize-observer-6FaCDon1.js | 18.37 kB | 0%
static/js/BackgroundImage-hmXOL_KY.js | 122.29 kB | 0%"
`;

exports[`displays module information when files are added/removed/changed 1`] = `
"
<details>
<summary>Changeset</summary>

File | Î” | Size
---- | - | ----
\`node_modules/actions/src/client/this-file-was-added.ts\` | ðŸ†• +1.46 kB | 0 B â†’ 1.46 kB
\`node_modules/actions/src/client/this-file-grew-larger.tsx\` | ðŸ“ˆ +200 B (+35.59%) | 562 B â†’ 762 B
\`node_modules/actions/src/client/helpers/this-file-grew-smaller.ts\` | ðŸ“‰ -200 B (-7.94%) | 2.46 kB â†’ 2.26 kB
\`node_modules/actions/src/client/this-file-will-be-deleted.ts\` | ðŸ”¥ -1.46 kB (-100%) | 1.46 kB â†’ 0 B

</details>
"
`;

exports[`displays no module information when unchanged 1`] = `
"
Changeset

No files were changed"
`;
